### PR TITLE
Recoverable tasks [version 1]

### PR DESCRIPTION
**Pull request by acierto**
## Hey, I just made a Pull Request!

That's a slice of this https://github.com/backstage/backstage/pull/21299.
Included the logic to restart the task from the beginning.
It makes the logic way more simple to create a good basis to build on top of it more smart way of recovering tasks after server crash or redeploy.
At this stage engineer is responsible to make sure that the task is fully idempotent and can run all steps once again.
It means that at this moment you can't use standard actions out-of-the-box, but come up with your own solution

Closed old PR https://github.com/backstage/backstage/pull/21814  cause of DCO issues which is not possible to solve without affecting the reviewers. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by benjdlambert — COMMENTED** [review]
Probably gonna look at merging this after the releases that's going out today I think. Need to have a deeper look but saw this that we called out in the other PR first.

File: docs/features/techdocs/creating-and-publishing.md
Comment (benjdlambert) [inline docs/features/techdocs/creating-and-publishing.md:1]: Maybe undo these changes from this PR 👍

File: plugins/scaffolder-backend/api-report.md [hunk 1]
```diff
@@ -552,6 +559,8 @@ export interface TaskStore {
     }[];
   }>;
   // (undocumented)
+  recoverTasks?(options: TaskStoreRecoverTaskOptions): Promise<string[]>;
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/api-report.md:562]: Return type: `Promise<{ ids: string[]}>` so that we can extend it easily in the future.

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 2]
```diff
@@ -223,34 +240,31 @@ export class DatabaseTaskStore implements TaskStore {
...
-          status: 'processing',
-          lastHeartbeatAt: task.last_heartbeat_at,
-          createdAt: task.created_at,
-          createdBy: task.created_by ?? undefined,
-          secrets,
-        };
-      } catch (error) {
-        throw new Error(`Failed to parse spec of task '${task.id}', ${error}`);
-      }
+      const secrets = task.secrets ? JSON.parse(task.secrets) : undefined;
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:258]: This wouldn't be caught like it was previously, maybe something to think about.

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 3]
```diff
@@ -223,34 +240,31 @@ export class DatabaseTaskStore implements TaskStore {
...
+      return {
+        id: task.id,
+        spec,
+        status: 'processing',
+        lastHeartbeatAt: task.last_heartbeat_at,
+        createdAt: task.created_at,
+        createdBy: task.created_by ?? undefined,
+        secrets,
+      } as SerializedTask;
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:267]: Why is this cast needed?

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 4]
```diff
@@ -223,34 +240,31 @@ export class DatabaseTaskStore implements TaskStore {
...
-          secrets,
-        };
-      } catch (error) {
-        throw new Error(`Failed to parse spec of task '${task.id}', ${error}`);
-      }
+      const secrets = task.secrets ? JSON.parse(task.secrets) : undefined;
+      return {
+        id: task.id,
+        spec,
+        status: 'processing',
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:262]: `'processing' as TaskStatus`?

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 5]
```diff
@@ -462,4 +479,52 @@ export class DatabaseTaskStore implements TaskStore {
...
+      );
+      if (this.db.client.config.client.includes('mysql')) {
+        heartbeatInterval = this.db.raw(
+          `date_sub(now(), interval ${timeoutS} second)`,
+        );
+      } else if (this.db.client.config.client.includes('sqlite3')) {
+        heartbeatInterval = this.db.raw(`datetime('now', ?)`, [
+          `-${timeoutS} seconds`,
+        ]);
+      }
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:500]: Break this out of this transaction for readability sake. The catalog also has some examples.

//freben

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 6]
```diff
@@ -415,6 +418,7 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
 
       const context: TemplateContext = {
         parameters: task.spec.parameters,
+        EXPERIMENTAL_recovery: task.spec.EXPERIMENTAL_recovery,
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:421]: Maybe this is not needed right now, let's check

File: plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts [hunk 7]
```diff
@@ -202,6 +205,32 @@ export class StorageTaskBroker implements TaskBroker {
...
+          timeout: readDuration(
+            this.config,
+            'scaffolder.EXPERIMENTAL_recoverTasksTimeout',
+            {
+              seconds: 30,
+            },
+          ),
+        })) ?? [];
+      recoveredTaskIds.forEach(() => {
+        this.signalDispatch();
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts:228]: Think we only need to do this once, as it's only affecting the current node. The other nodes will eventually run this on their own when there's a new task.

File: plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts [hunk 8]
```diff
@@ -202,6 +205,32 @@ export class StorageTaskBroker implements TaskBroker {
...
+  public async recoverTasks(): Promise<boolean> {
+    const enabled =
+      (this.config &&
+        this.config.getOptionalBoolean(
+          'scaffolder.EXPERIMENTAL_recoverTasks',
+        )) ??
+      false;
+
+    if (enabled) {
+      const recoveredTaskIds =
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts:217]: Maybe let's shallow this nesting a little bit and break it out into mutliple lines for readablility.

File: plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts [hunk 9]
```diff
@@ -111,7 +112,21 @@ export class TaskWorker {
     });
   }
 
+  async recoverTasks() {
+    try {
+      await this.options.taskBroker.recoverTasks?.();
+    } catch (_err) {
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:118]: Should we at least log some things in here? Just for observability.

File: plugins/scaffolder-backend/src/scaffolder/tasks/taskRecoveryHelper.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SerializedTaskEvent } from '@backstage/plugin-scaffolder-node';
+import { TaskRecoverStrategy } from '@backstage/plugin-scaffolder-common';
+
+export const compactEvents = (
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/taskRecoveryHelper.ts:20]: `trimPreviousAttemptEvents`

Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/tasks/taskRecoveryHelper.ts:20]: Maybe something better?

File: plugins/scaffolder-backend/src/service/router.ts [hunk 11]
```diff
@@ -331,7 +333,14 @@ export async function createRouter(
       });
 
   actionsToRegister.forEach(action => actionRegistry.register(action));
-  workers.forEach(worker => worker.start());
+
+  const launchWorkers = () => workers.forEach(worker => worker.start());
+
+  if (options.lifecycle) {
+    options.lifecycle.addStartupHook(launchWorkers);
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/service/router.ts:340]: Maybe `worker.stop` in `addShutdownHook` too?

**Review by benjdlambert — APPROVED** [review]
Nice! Let's merge this an iterate! :tada:

**Comment by Alok650** [comment]
Hi @benjdlambert currently tasks are started from the first stage when use start-over to restart tasks with same configs. However, we have use cases wherein we want to start from a particular stage i.e. the stage which has failed. Is it possible to do? If not, what would be an efficient way to do this?

**Comment by benjdlambert** [comment]
@Alok650 this is on the radar, and something that we will be picking up - it's dependant on workspace serialization and things though first.
